apps/emqx/src/emqx_channel.erl

```diff
@@ -2099,9 +2118,19 @@ ensure_disconnected(
 %%--------------------------------------------------------------------
 %% Maybe Publish will msg
 -spec maybe_publish_will_msg(Reason, channel()) -> channel() when
-    Reason :: kick | sock_closed | {shutdown, atom()}.
+    Reason :: takenover | kicked | discarded | expired | sock_closed | {shutdown, atom()}.
```

Comment:
Q: Why there's no `not_authorized` that's being handled in one of the following clauses?

Reply:
Now I removed it, look unnecessary now.